xformers/components/attention/csrc/cuda/mem_eff_attention/attention_backward_generic.cu

```diff
@@ -211,6 +230,69 @@ mem_efficient_attention_backward_cutlass(
     ASSIGN_CHECK_OVERFLOW(p.k_strideH, key.stride(2));
     ASSIGN_CHECK_OVERFLOW(p.v_strideH, value.stride(2));
 
+    if (bias.has_value()) {
+      CHECK_NOSPARSE_LASTCONTIGUOUS_CUDA((*bias));
+
+      p.bias_ptr = (scalar_t*) bias->data_ptr();
+
+      // assign strides for both B and gB. gB might have different strides than
+      // B, for example if bias tensor was created with
+      // torch.tensor((B * nH, 1, nK)).expand((B * nH, nQ, nK)),
+      // different values of Q will point to the same memory locations,
+      // meaning bias.stride(1) == 0
+      if (bias->dim() == 2) {  // (n_queries, n_keys)
+        TORCH_INTERNAL_ASSERT(bias->size(0) == M);
+        TORCH_INTERNAL_ASSERT(bias->size(1) == N);
+
+        ASSIGN_CHECK_OVERFLOW(p.bias_strideB, 0);
+        ASSIGN_CHECK_OVERFLOW(p.bias_strideH, 0);
+        ASSIGN_CHECK_OVERFLOW(p.bias_strideM, grad_bias.stride(0));
+
+        if (bias_requires_grad) {
+          ASSIGN_CHECK_OVERFLOW(p.gB_strideB, 0);
+          ASSIGN_CHECK_OVERFLOW(p.gB_strideH, 0);
+          ASSIGN_CHECK_OVERFLOW(p.gB_strideM, bias->stride(0));
+        }
+      } else if (bias->dim() == 3) {  // (batch_sz * n_heads, n_queries, n_keys)
+        TORCH_INTERNAL_ASSERT(bias->size(0) == B * nH);
+        TORCH_INTERNAL_ASSERT(bias->size(1) == M);
+        TORCH_INTERNAL_ASSERT(bias->size(2) == N);
+
+        ASSIGN_CHECK_OVERFLOW(p.bias_strideB, nH * bias->stride(0));
+        ASSIGN_CHECK_OVERFLOW(p.bias_strideH, bias->stride(0));
+        ASSIGN_CHECK_OVERFLOW(p.bias_strideM, bias->stride(1));
+
+        if (bias_requires_grad) {
+          ASSIGN_CHECK_OVERFLOW(p.gB_strideB, nH * grad_bias.stride(0));
+          ASSIGN_CHECK_OVERFLOW(p.gB_strideH, grad_bias.stride(0));
+          ASSIGN_CHECK_OVERFLOW(p.gB_strideM, grad_bias.stride(1));
+        }
```

Comment:
awesome, will do

Reply:
qq: [the tests use 3 dims](https://github.com/facebookresearch/xformers/blob/main/tests/test_mem_eff_attention.py#L233-L234). would you prefer i change the tests (not sure if other implementations can handle 4 dims), or handle cases for 3 and 4?

Reply:
Let's only support dim=4 in the kernel, and expand properly in python before we call the C++ code.
Later we can simplify that to accept masks with or without head dimension (because we accept q/k/v in both BMHK and BMK shapes, so it makes sense to be coherent for the mask as well)